api/client/websocketupgradeproto/protocol.go

```diff
@@ -0,0 +1,570 @@
+			// that we know will never come.
+			if !c.supportsCloseProccess {
+				unblockReaderOnErr()
+				return
+			}
+
+			// Set a read deadline to avoid blocking forever waiting for
+			// the close frame from the remote side that may never come.
+			const deadline = 3 * time.Second
+			c.underlyingConn.SetReadDeadline(time.Now().Add(deadline))
+
+			// Once we write the close frame, we can return from this loop.
+			// We will never need to write anything else into the connection,
+			// so we can just wait for the read loop to read the close frame
+			// from the remote side and close the connection.
+			return
+		}
+		c.cond.Wait()
+	}
+}
+
+// writeFrameLocked writes a WebSocket frame to the connection without acquiring
+// the write mutex. This is used when we already hold the write mutex.
+func (c *Conn) writeFrame(frame ws.Frame) error {
+	// If the connection is a client connection, we should mask the frame
+	// as per the WebSocket protocol. In this case, we use a empty mask
+	// for simplicity so messages are not actually masked, but the masked bit is set.
+	frame.Header.Masked = c.connType == clientConnection
+
+	// Unlock the mutex while writing to avoid blocking other operations.
+	// The mutex is re-locked after the write is complete.
+	c.mu.Unlock()
+	defer c.mu.Lock()
+	// There is no need to mask from server to client.
+	return ws.WriteFrame(c.underlyingConn, frame)
+}
+
+// WritePing sends a Ping frame to the client.
+func (c *Conn) WritePing() error {
+	// Clients never send pings.
+	if c.connType == clientConnection {
+		return nil
+	}
+
+	// Create a Ping frame with the Teleport component as payload.
+	pingFrame := ws.NewPingFrame([]byte(ComponentTeleport))
+
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	// Queue the ping frame to be sent in the next write loop iteration
+	// and signal the write loop.
+	c.pingPongReplies = append(c.pingPongReplies, pingFrame)
+	c.cond.Broadcast()
+	return nil
+}
+
+// Protocol returns the negotiated WebSocket sub-protocol for this connection.
+func (c *Conn) Protocol() string {
+	return c.protocol
+}
+
+// Close closes the WebSocket connection gracefully by sending a close frame
+// to the remote side if supported.
+func (c *Conn) Close() error {
+	closeFrame := ws.NewCloseFrame(
+		ws.NewCloseFrameBody(ws.StatusNormalClosure, ""),
+	)
+	return c.closeWithErrFrame(closeFrame)
+}
+
+func (c *Conn) closeWithErrFrame(frame ws.Frame) error {
+	// This afterFunc sets a deadline to unblock the read and write loops in case
+	// the remote side never acknowledges previously sent frames and the
+	// write loop gets stuck trying to send any frame.
+	// If the remote side does not respond, the deadline will trigger and
+	// the read and write loops will exit. In such cases, we don't properly
+	// complete the close handshake, but connection wasn't healthy anyway.
+	stopTimer := time.AfterFunc(
+		maxDeadlineOnClose,
+		func() {
+			c.underlyingConn.SetDeadline(time.Now().Add(-time.Second))
+		},
```

Comment:
I'd just close the underlying connection after the timeout, it's not like we need the nondestructive unblocking if the connection was closed and we did our best to send a close frame to the remote side.